modules/wallet/scan.go

```diff
@@ -0,0 +1,198 @@
+type scannedOutput struct {
+	id        types.OutputID
+	value     types.Currency
+	seedIndex uint64
+}
+
+// A seedScanner scans the blockchain for addresses that belong to a given
+// seed.
+type seedScanner struct {
+	blockheight      types.BlockHeight
+	dustThreshold    types.Currency              // minimum value of outputs to be included
+	keys             map[types.UnlockHash]uint64 // map address to seed index
+	largestIndexSeen uint64                      // largest index that has appeared in the blockchain
+	minerOutputs     map[types.BlockHeight][]scannedOutput
+	seed             modules.Seed
+	siacoinOutputs   map[types.OutputID]scannedOutput
+}
+
+func (s *seedScanner) numKeys() uint64 {
+	return uint64(len(s.keys))
+}
+
+// generateKeys generates n additional keys from the seedScanner's seed.
+func (s *seedScanner) generateKeys(n uint64) {
+	initialProgress := s.numKeys()
+	for i, k := range generateKeys(s.seed, initialProgress, n) {
+		s.keys[k.UnlockConditions.UnlockHash()] = initialProgress + uint64(i)
+	}
+}
+
+// ProcessConsensusChange scans the blockchain for information relevant to the
+// seedScanner.
+func (s *seedScanner) ProcessConsensusChange(cc modules.ConsensusChange) {
+	s.blockheight -= types.BlockHeight(len(cc.RevertedBlocks))
+	// update outputs
+	for _, block := range cc.AppliedBlocks {
+		s.blockheight++
+		// when a miner payout matures, move it from minerOutputs to
+		// siacoinOutputs
+		if matured, exists := s.minerOutputs[s.blockheight]; exists {
+			for _, maturedOutput := range matured {
+				s.siacoinOutputs[maturedOutput.id] = maturedOutput
+			}
+			delete(s.minerOutputs, s.blockheight)
+		}
+		for i, mp := range block.MinerPayouts {
+			// if a seed miner output is found, add it to the delayed output
+			// set
+			if index, exists := s.keys[mp.UnlockHash]; exists && mp.Value.Cmp(s.dustThreshold) > 0 {
+				maturityHeight := s.blockheight + types.MaturityDelay
+				s.minerOutputs[maturityHeight] = append(s.minerOutputs[maturityHeight], scannedOutput{
+					id:        types.OutputID(block.MinerPayoutID(uint64(i))),
+					value:     mp.Value,
+					seedIndex: index,
+				})
+			}
+		}
+		for _, txn := range block.Transactions {
+			for i, sco := range txn.SiacoinOutputs {
+				// if a seed output is found, add it to the output set
+				if index, exists := s.keys[sco.UnlockHash]; exists && sco.Value.Cmp(s.dustThreshold) > 0 {
+					id := types.OutputID(txn.SiacoinOutputID(uint64(i)))
+					s.siacoinOutputs[id] = scannedOutput{
+						id:        id,
+						value:     sco.Value,
+						seedIndex: index,
+					}
+				}
+			}
+			// TODO: should this be done in a separate txn loop? This would be
+			// mandatory if txn0 can spend an output created in txn1
+			for _, sci := range txn.SiacoinInputs {
+				// if a seed output is spent, remove it from the output set
+				id := types.OutputID(sci.ParentID)
+				if _, exists := s.siacoinOutputs[id]; exists {
+					delete(s.siacoinOutputs, id)
+				}
+			}
+		}
+	}
```

Comment:
I don't think you need to be looking at the actual blocks to achieve this, I also don't think you need to be tracking the miner outputs yourself. Because doing this stuff has a lot of surface area for bugs, it's all handled in the ConsensusChange object for you. You can just iterate through `cc.SiacoinOutputDiffs` and `cc.SiafundOutputDiffs`, ignoring the blocks altogether.